src/jdk.jdwp.agent/share/native/libjdwp/util.c

```diff
@@ -1099,31 +1106,394 @@ debugMonitorNotifyAll(jrawMonitorID monitor)
+    JDI_ASSERT(dbg_monitor->monitor == NULL);
 
     error = JVMTI_FUNC_PTR(gdata->jvmti,CreateRawMonitor)
                 (gdata->jvmti, name, &monitor);
     if (error != JVMTI_ERROR_NONE) {
         EXIT_ERROR(error, "on creation of a raw monitor");
     }
-    return monitor;
+
+    JDI_ASSERT(monitor != NULL);
+
+    // Need to lock during initialization so verifyMonitorRank() can be guaranteed that
+    // if the monitor field is set, then the monitor if fully initialized. More specifically,
+    // that the rank field has been set.
+    if (gdata->rankedMonitors) {
+        dbgRawMonitor_lock();
+    }
+    dbg_monitor->monitor = monitor;
+    dbg_monitor->rank = dbg_monitor_rank;
+    dbg_monitor->name = name;
+    if (gdata->rankedMonitors) {
+        dbgRawMonitor_unlock();
+    }
+
+    return dbg_monitor;
 }
 
 void
-debugMonitorDestroy(jrawMonitorID monitor)
+debugMonitorDestroy(DebugRawMonitor *dbg_monitor)
 {
     jvmtiError error;
+    JDI_ASSERT(dbg_monitor != NULL);
 
     error = JVMTI_FUNC_PTR(gdata->jvmti,DestroyRawMonitor)
-                (gdata->jvmti, monitor);
+                (gdata->jvmti, dbg_monitor->monitor);
     error = ignore_vm_death(error);
     if (error != JVMTI_ERROR_NONE) {
         EXIT_ERROR(error, "on destruction of raw monitor");
     }
+
+    dbg_monitor->monitor = NULL;
+}
+
+static char*
+getThreadName(jthread thread)
+{
+    jvmtiThreadInfo info;
+    jvmtiError error;
+
+    memset(&info, 0, sizeof(info));
+    error = JVMTI_FUNC_PTR(gdata->jvmti,GetThreadInfo)
+            (gdata->jvmti, thread, &info);
+    return info.name;
+}
+
+void
+dumpRawMonitor(DebugRawMonitor *dbg_monitor) {
+    tty_message("DebugRawMonitor(%d)", dbg_monitor->rank);
+    tty_message("\tname: %s", dbg_monitor->name);
+    if (dbg_monitor->ownerThread != NULL) {
+        char* threadName = getThreadName(dbg_monitor->ownerThread);
+        tty_message("\townerThreadName: %s", threadName);
+        jvmtiDeallocate(threadName);
+    }
+}
+
+void
+dumpRawMonitors() {
+    int i;
+    for (i = 0; i < NUM_DEBUG_RAW_MONITORS; i++) {
+        DebugRawMonitor* dbg_monitor = &dbg_monitors[i];
+        if (dbg_monitor == NULL) {
+            continue;
+        }
+        jthread thread = dbg_monitor->ownerThread;
+        if (thread == NULL) {
+            continue;
+        }
+        dumpRawMonitor(dbg_monitor);
+    }
+}
+
+static void
+assertIsCurrentThread(JNIEnv *env, jthread thread)
+{
+    if (gdata->vmDead) {
+        return; // This assert is not reliable if the VM is exiting
+    }
+    jthread current_thread = threadControl_currentThread();
+    if (!isSameObject(env, thread, current_thread)) {
+        tty_message("ERROR: Threads not the same: %p %p\n", thread, current_thread);
+        dumpRawMonitors();
+        JDI_ASSERT(JNI_FALSE);
+    }
+    JNI_FUNC_PTR(env,DeleteLocalRef)(env, current_thread);
+}
+
+static void
+verifyMonitorRank(JNIEnv *env, DebugRawMonitorRank rank, jthread thread)
```

Comment:
please add a comment that the function should be called under dbgRawMonitor lock (also for other functions which assume this)

Reply:
ok. I was also thinking of setting a static that contains the jthread of the thread that holds the lock, and we can assert it is held.

Reply:
Fixed.